src/python/pants/backend/go/util_rules/build_opts.py

```diff
@@ -222,17 +240,68 @@ async def go_extract_build_options_from_target(
         )
         with_msan = False
 
+    # Extract the `with_asan` value for this target.
+    with_asan, asan_reason = _first_non_none_value(
+        [
+            (
+                True if go_test_subsystem.force_asan and test_target_fields else None,
+                "the `[go-test].force_asan` option is in effect",
+            ),
+            (
+                test_target_fields.test_asan.value,
+                f"{GoTestAddressSanitizerEnabledField.alias}={test_target_fields.test_asan.value} on target `{request.address}`",
+            )
+            if test_target_fields
+            else None,
+            (
+                target_fields.asan.value,
+                f"{GoAddressSanitizerEnabledField.alias}={target_fields.asan.value} on target `{request.address}`",
+            )
+            if target_fields
+            else None,
+            (
+                go_mod_target_fields.asan.value,
+                f"{GoAddressSanitizerEnabledField.alias}={go_mod_target_fields.asan.value} on target `{request.address}`",
+            )
+            if go_mod_target_fields
+            else None,
+            (False, "default"),
+        ]
+    )
+    if with_asan and not asan_supported(goroot):
+        logger.warning(
+            f"Interoperation with the C/C++ address sanitizer would have been enabled for target `{request.address}` "
+            f"because {asan_reason}, "
+            f"but the address sanitizer is not supported on platform {goroot.goos}/{goroot.goarch}."
+        )
+        with_asan = False
+
+    # Ensure that only one of the race detector, memory sanitizer, and address sanitizer are ever enabled
+    # at a single time.
     if with_race_detector and with_msan:
         raise ValueError(
-            "The Go data race detector and C/C++ memory sanitizer support cannot be enabled at the same time. "
+            "The Go data race detector and C/C++ memory sanitizer cannot be enabled at the same time. "
+            f"The Go data race detector is enabled because {race_detector_reason}. "
+            f"The C/C++ memory sanitizer is enabled because {msan_reason}."
+        )
+    if with_race_detector and with_asan:
+        raise ValueError(
+            "The Go data race detector and C/C++ address sanitizer cannot be enabled at the same time. "
             f"The Go data race detector is enabled because {race_detector_reason}. "
-            f"The C/C++ memory sanitizer support is enabled because {msan_reason}."
+            f"The C/C++ address sanitizer is enabled because {asan_reason}."
+        )
+    if with_msan and with_asan:
+        raise ValueError(
+            "The C/C++ memory and address sanitizers cannot be enabled at the same time. "
+            f"The C/C++ memory sanitizer is enabled because {msan_reason}. "
+            f"The C/C++ address sanitizer is enabled because {asan_reason}."
```

Comment:
These are some of the best error messages we have in the entire system!

Reply:
There's not going to be a 4th thing though is there? :)

Reply:
Thankfully not! This code block adapts https://github.com/golang/go/blob/d5de62df152baf4de6e9fe81933319b86fd95ae4/src/cmd/compile/internal/base/flag.go#L250-L257 which only deals with those three options.